### PR TITLE
merged and reduce finish added

### PR DESCRIPTION
**Pull request by vibhatha**

### CODE REVIEW

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/reduce/ReduceBatchReceiver.java [hunk 1]
```diff
@@ -116,23 +115,27 @@ public boolean onMessage(int source, int path, int target, int flags, Object obj
       if (object instanceof ChannelMessage) {
```
Comment (supunkamburugamuve) [inline twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/reduce/ReduceBatchReceiver.java]: I think we need to check for empty flag here

File: twister2/examples/src/java/edu/iu/dsc/tws/examples/internal/task/batch/GatherBatchTask.java [hunk 2]
```diff
@@ -99,10 +99,16 @@ public void init(Config config, int workerID, AllocatedResources resources,
...
     @Override
     public void run() {
-      ctx.write("gather-edge", "1");
+      count++;
+      if (count == 10) {
+        ctx.writeEnd("gather-edge", "Hello");
+      } else if (count < 10) {
+        ctx.write("gather-edge", "Hello");
+      }
     }
```
Comment (supunkamburugamuve) [inline twister2/examples/src/java/edu/iu/dsc/tws/examples/internal/task/batch/GatherBatchTask.java]: lets remove these from this pull rquest as we are only working on reduce

File: twister2/examples/src/java/edu/iu/dsc/tws/examples/internal/task/batch/PartitionBatchTask.java [hunk 3]
```diff
@@ -88,9 +88,9 @@ public void init(Config config, int workerID, AllocatedResources resources,
     @Override
     public void run() {
       count++;
-      if (count == 1000) {
+      if (count == 20) {
         ctx.writeEnd("partition-edge", "Hello");
-      } else if (count < 1000) {
+      } else if (count < 20) {
         ctx.write("partition-edge", "Hello");
```
Comment (supunkamburugamuve) [inline twister2/examples/src/java/edu/iu/dsc/tws/examples/internal/task/batch/PartitionBatchTask.java]: lets remove these from this pull rquest as we are only working on reduce

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/api/ParallelOperationFactory.java [hunk 4]
```diff
@@ -86,6 +86,7 @@ public IParallelOperation build(Edge edge, Set<Integer> sources, Set<Integer> de
               edge.getName());
           return bcastOp;
         } else if (CommunicationOperationType.BATCH_GATHER.equals(edge.getOperation())) {
+          System.out.println("Gather Ops");
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/api/ParallelOperationFactory.java]: Remove the sout

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/AllReduceBatchOperation.java [hunk 5]
```diff
@@ -40,7 +40,7 @@ public AllReduceBatchOperation(Config config, Communicator network, TaskPlan tPl
   }
 
   public void prepare(Set<Integer> sources, Set<Integer>  dest, EdgeGenerator e,
-                      DataType dataType, String edgeName) {
+                      MessageType dataType, String edgeName) {
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/AllReduceBatchOperation.java]: lets not use message type

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/ReduceBatchOperation.java [hunk 6]
```diff
@@ -91,21 +98,21 @@ public boolean onMessage(int source, int path, int target, int flags, Object obj
...
-      return true;
+      return true; //op.progress();
     }
 
     @Override
     public boolean receive(int target, Object object) {
       TaskMessage msg = new TaskMessage(object,
           edge.getStringMapping(communicationEdge), target);
       outMessages.get(target).offer(msg);
+      System.out.println(msg.getContent().getClass());
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/ReduceBatchOperation.java]: remove sout

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java [hunk 7]
```diff
@@ -148,14 +149,15 @@ public boolean execute() {
       if (isDone) {
         state.set(InstanceState.EXECUTION_DONE);
       }
-
+      int flag = 0;
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java]: we dont need flag

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java [hunk 8]
```diff
@@ -214,4 +217,16 @@ public void registerOutParallelOperation(String edge, IParallelOperation op) {
   public int getWorkerId() {
     return workerId;
   }
+
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java]: we don't need this method

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/threading/BatchSharingExecutor.java [hunk 9]
```diff
@@ -58,6 +58,7 @@ public boolean runExecution() {
       try {
         t.join();
       } catch (InterruptedException e) {
+        e.printStackTrace();
```
Comment (supunkamburugamuve) [inline twister2/executor/src/java/edu/iu/dsc/tws/executor/threading/BatchSharingExecutor.java]: printStackTrace

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/api/ParallelOperationFactory.java (hunk 4 above)
Comment (vibhatha) [reply twister2/executor/src/java/edu/iu/dsc/tws/executor/api/ParallelOperationFactory.java]: removed and will add to next commit.

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/ReduceBatchOperation.java (hunk 6 above)
Comment (vibhatha) [reply twister2/executor/src/java/edu/iu/dsc/tws/executor/comm/operations/batch/ReduceBatchOperation.java]: removed and will add to next commit

File: twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java (hunk 7 above)
Comment (vibhatha) [reply twister2/executor/src/java/edu/iu/dsc/tws/executor/core/batch/SourceBatchInstance.java]: removed and next commit will include it.

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/reduce/ReduceBatchReceiver.java (hunk 1 above)
Comment (vibhatha) [reply twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/reduce/ReduceBatchReceiver.java]: not clear to me. need to discuss.

**Pull request closed without merging**